pyxem/signals/common_diffraction.py

```diff
@@ -32,6 +36,50 @@
 class CommonDiffraction:
     """Common functions for all Diffraction Signals classes"""
 
+    def to_device(self):  # pragma: no cover
+        """Transfer the data to the GPU.
+
+        Returns
+        -------
+        Diffraction2D
+            The data on the GPU.
+        """
+        if not CUPY_INSTALLED:
+            raise ImportError(
+                "The cupy package is required to use this method. "
+                "Please install it using `conda install cupy`."
+            )
+        if not self._gpu:
+            if self._lazy:
+                self.data = self.data.map_blocks(
+                    cp.asarray
+                )  # pass chunk-wise the data to GPU
+            else:
+                self.data = cp.asarray(self.data)  # pass all the data to the GPU
+        self._gpu = True
+
+    def to_host(self):  # pragma: no cover
+        """Transfer the data from the GPU to the CPU."""
+        if not CUPY_INSTALLED:
+            raise ImportError(
+                "The cupy package is required to use this method. "
+                "Please install it using `conda install cupy`."
+            )
+        if self._gpu:
+            if self._lazy:
+                self.data = self.data.map_blocks(cp.asnumpy)
+            else:
+                self.data = cp.asnumpy(self.data)
+        self._gpu = False
+
+    @property
+    def _gpu(self):
+        self.metadata.get_item("_HyperSpy.gpu")
```

Comment:
Why does it need to be saved in the metadata? This would be saved to file, which will be incorrect when loading the file later one?

Reply:
It's a problem when you change from one signal type to another using the set_signal_type attributes of the signal are lost.  It's kind of a hacky way of getting around this. 

If the only way to save the file  is with the cpu it shouldn't be an issue. That being said if you get the direct i-o with a gpu to work, it no longer will but in that case you should be able to handle that with rosettasciio and overwrite that metadata. 

Maybe we have something like "temporary" metadata storage which isn't saved but exists for as long as the object does and can pass from one instance to the next.

Reply:
> It's a problem when you change from one signal type to another using the set_signal_type attributes of the signal are lost. It's kind of a hacky way of getting around this.

`set_signal_type` keep the same object and therefore all attributes:
```python
import hyperspy.api as hs

s = hs.data.two_gaussians()
print(id(s))
assert s.metadata.Signal.signal_type == ""
s.test = True
s.set_signal_type("EELS")
print(id(s))
assert s.test
assert s.metadata.Signal.signal_type == "EELS"
```

Maybe you mean when creating a new object when using something like `_deepcopy_with_new_data`? The same issue applies when creating a signal from a cupy array. If this is also set in the class constructor this should fix these two issues - with `_gpu` being simply an attribute and not saved in the metadata. Similarly to lazy signal, this is a runtime characteristic (as opposed to one of the data itself) and it doesn't feel right to be stored in the metadata.

An alternative would be to use the following approach https://github.com/hyperspy/hyperspy/blob/325f611a4323ab6e78793d735679a16f2ccd31dc/hyperspy/misc/utils.py#L1540-L1560.

However, depending on where it is used, it can add non-negligible overhead.

Reply:
> Maybe you mean when creating a new object when using something like `_deepcopy_with_new_data`? The same issue applies when creating a signal from a cupy array. If this is also set in the class constructor this should fix these two issues - with `_gpu` being simply an attribute and not saved in the metadata. Similarly to lazy signal, this is a runtime characteristic (as opposed to one of the data itself) and it doesn't feel right to be stored in the metadata.

You're correct, I do mean `_deepcopy_with_new_data` I never tracked the logic backwards and just assumed that the `set_signal_type` was causing the issue.  I don't think checking the array type is a great idea.  I don't love checking the underlying dask array type as you would have to compute the array and while it's usually fairly fast it can be slow/ it forces you to run serially rather than in parallel.  I'd rather not take the performance hit, especially for something like GPU computing where the reason for using it is speed.

Reply:
Actually, I think `dask` is better at keeping track of this so now it should be accessible without computing!

Reply:
Yes, there is the `_meta` attributes of dask array but it is private. Maybe https://docs.dask.org/en/stable/generated/dask.array.utils.meta_from_array.html#dask.array.utils.meta_from_array?